Move protocol-specific user presence checking code from Env to CTAP library

egor-duda opened the pull request:
Cleaned-up version of #497, part of implementation for #485. Code changes in this PR:
* Add `CtapDuration` type for timeouts in CTAP library
* Add `SendOrRecvStatus` and `SendOrRecvResult` types for results of I/O operations between authenticator and UA
* Add `UserPresenceStatus` and `UserPresenceResult` types for results of user presence checks
* Add main & vendor HID I/O channels to Env, use them for sending data from authenticator to UA
* Move `send_keepalive_up_needed` and `check_user_presence` functions from TockEnv to CTAP library
* Change `UserPresence` trait to provide fine-grained functions for user presence checking, along with `quick_check` function for short-circuiting user presence check for testing purposes

egor-duda committed:
Common duration type for ctap library independent of TockOS

egor-duda committed:
Implement Env-specific ctap-hid channels for I/O
Common I/O Status, Error and Result types

egor-duda committed:
Move common user presence checking code to ctap library

ia0 commented on src/env/mod.rs:
```diff
@@ -10,11 +11,57 @@ use rng256::Rng256;
+    Timeout,
+}
+
+pub type UserPresenceResult = Result<UserPresenceStatus, Ctap2StatusCode>;
```
FYI: Ideally we should not use `Ctap2StatusCode` but specific error-types instead. The `Env` implementer should not need to know about CTAP. The CTAP library would convert from the `Env` interface error type to the appropriate `Ctap2StatusCode`. But as a first step I think it's fine.

kaczmarczyck replied:
Where do we use the `Ctap2StatusCode`? Isn't the `UserPresenceStatus` translated into `Ctap2StatusCode` later? Why not just return `UserPresenceStatus` directly?

Random idea: We could also make it a Result<(), TimeoutOrCanceledOrDeclined> and implement `From` for `Ctap2StatusCode` for simple conversion on the call site of `wait_for_user_presence_with_timeout`.

ia0 commented on src/env/mod.rs:
```diff
@@ -10,11 +11,57 @@ use rng256::Rng256;
 pub trait UserPresence {
-    /// Blocks for user presence.
+    /// Called at the beginning of user presence checking process.
+    fn user_presence_check_init(&mut self, _channel: Channel) {}
```
I would not repeat `user_presence` in those methods. This is redundant: `env.user_presence().user_presence_check_init()`. I would name them:

```
fn check_init()
fn wait_with_timeout()
fn check_complete()
```

ia0 commented on src/env/mod.rs:
```diff
@@ -10,11 +11,57 @@ use rng256::Rng256;
+        _channel: Channel,
+        _timeout: CtapDuration,
+    ) -> UserPresenceResult {
+        Ok(UserPresenceStatus::Confirmed)
```
I don't think we should provide default implementation for any of those functions. Default implementations are useful when they is always correct. They can be overwritten for performance reasons only. Otherwise it's too easy to forget to implement it and get an incorrect implementation.

ia0 commented on src/env/mod.rs:
```diff
@@ -10,11 +11,57 @@ use rng256::Rng256;
+    ) -> SendOrRecvResult;
+}
+
+pub enum UserPresenceStatus {
```
nit: We could move all user presence related stuff to its own `api::user_presence` module, to keep this module simple and only about `Env`.

ia0 commented on src/clock.rs:
```diff
@@ -57,6 +57,7 @@ pub fn new_clock() -> CtapClock {
 }
 
 pub type CtapInstant = embedded_time::Instant<CtapClock>;
+pub type CtapDuration = Milliseconds<ClockInt>;
```
I'm not sure this alias is beneficial because it hides that it's milliseconds.

egor-duda replied:
I was looking at `libtock_drivers::timer::Duration` as a template for naming.
Should this type be renamed into something like CtapDurationMs, or just replaced everywhere with `embedded_time::duration::Milliseconds<ClockInt>` ?

kaczmarczyck replied:
Use `embedded_time::duration::Milliseconds<ClockInt>` for now please.

ia0 commented on src/lib.rs:
```diff
@@ -118,4 +118,13 @@ impl<E: Env> Ctap<E> {
         self.hid.update_wink_timeout(now);
     }
+
+    pub fn main_hid_channel(&mut self) -> &mut E::CtapHidChannel {
```
I'd say, let's remove the `std` feature-gate of `env()` and access the HID channels from there. This permits to keep the API simple until we get a better idea of whether accessing `Env` is meaningful in prod (but I start to suspect yes).

ia0 commented on src/main.rs:
```diff
@@ -132,26 +134,29 @@ fn main() {
-                        .flex_unwrap();
-                match status {
-                    usb_ctap_hid::SendOrRecvStatus::Timeout => {
+                let channel = match transport {
```
I suspect this `match` to be useful more than once. Let's make this a method of `Transport`:

```rust
/// Panics if the transport is not HID.
fn Transport::hid_channel<E: Env>(self, env: &mut E> -> &mut E::HidChannel;
```

ia0 commented on src/ctap/mod.rs:
```diff
@@ -244,6 +244,123 @@ fn verify_signature(
+    };
+    let keepalive_msg = CtapHid::keepalive(cid, KeepaliveStatus::UpNeeded);
+    for mut pkt in keepalive_msg {
+        let ctap_hid_channel = match transport {
```
This could reuse the `Transport::hid_channel` helper function.

kaczmarczyck replied:
Alternatively, we pass in the `ctap_hid_channel` and `cid`. Then we can have one `match` in `check_user_presence` that handles all transports.

ia0 commented on src/ctap/mod.rs:
```diff
@@ -244,6 +244,123 @@ fn verify_signature(
+    channel: Channel,
+    timeout: CtapDuration,
+) -> Result<(), Ctap2StatusCode> {
+    let (cid, transport) = match channel {
```
This should probably be a method of `Channel`:

```rust
fn Channel::split(self) -> (ChannelId, Transport);
```

kaczmarczyck replied:
NFC and BLE don't have a channel ID. This split is HID specific.

ia0 replied:
Right, wasn't sure, in that case:

```rust
/// Panics if not HID.
fn Channel::split_hid(self) -> (ChannelId, Transport);
```

kaczmarczyck replied:
I'm happy this way, and it may or may not move with NFC. Leaving it open for @ia0 .

ia0 replied:
If this is the only place we do this, SGTM.

ia0 commented on src/env/mod.rs:
```diff
@@ -10,11 +11,57 @@ use rng256::Rng256;
+    /// packets to the user agent. Implementations may use this function to override this behavior
+    /// and return Some(Result<...>) to pass it immediately to the caller wanting to check user
+    /// presence.
+    fn quick_check(&mut self, _channel: Channel) -> Option<Result<(), Ctap2StatusCode>> {
```
I'm not convinced we need this long-term. I see 2 options:
- It's possible to implement the test behavior with the init/wait/complete API. In which case we should remove this.
- It's not (yet) possible to implement the test behavior with the existing API. In which case we can keep this and explicitly mention that it's for tests and temporary.

ia0 commented on src/ctap/mod.rs:
```diff
@@ -244,6 +244,123 @@ fn verify_signature(
+    // First, send a keep-alive packet to notify that the keep-alive status has changed.
+    // All fallible functions are called without '?' operator to always reach
+    // user_presence_check_complete(...) cleanup function.
+    let mut result = send_keepalive_up_needed(env, channel, KEEPALIVE_DELAY)
```
Can't this whole code look like the following instead?

```rust
check_init();
let mut result = ...;
for i in 0..=TIMEOUT_ITERATIONS {
    result = send_keepalive();
    if result.is_err() || i == TIMEOUT_ITERATIONS { break; }
    result = wait_with_timeout();
    if result.is_err() { break; }
}
check_complete();
```
        
Essentially having only one call-site for `send_keepalive`. Unless I'm mistaken this should have the same behavior.

ia0 commented on src/env/tock/mod.rs:
```diff
@@ -18,10 +21,46 @@ use rng256::TockRng256;
+            {
+                Ok(SendOrRecvStatus::Received)
+            }
+            _ => Err(SendOrRecvError),
```
Interesting, if we're receiving something on the other channel, we're dropping it. This might be problematic. Not sure what's the best solution with the current design. @kaczmarczyck any ideas?

kaczmarczyck replied:
Could actually be okay. On a higher level, multiplexing isn't that good of an idea either, for reasons like:
- Stateful commands could drop each other's state in a livelock.
- Users have a hard time understanding what they touch for.

In this specific case, we are still responding. This is should have priority over accepting a new request. Also, the main loop handles receiving packets as well, so it can only occur if the packet arrived while the main loop / command was processed.

kaczmarczyck reviewed (commented):
I decided I want to write a high-level response first, before I look code details more. See all the details of this implementation, I noticed we might need some discussion. "UP" will be short for "user presence". Some of this might be obvious to everyone, but I still want to make it explicit:

First, let me start with a few goals of what we should be able to support:
- A user presence check over HID waits for a button touch while sending HID packets and blinking LEDs.
- A user presence check over HID during tests pretends that the button was pressed immediately.
- The CTAP library skips user presence checks since the transport of incoming packets was NFC.

Therefore, what is CTAP, what is Env, what is transport specific?
- Checking UP is a CTAP function, behavior depends on transport.
- How to receive UP over HID is Env specific. You could have an OpenSK that waits for a button, or one that doesn't even have buttons and accepts everything in the first 30 seconds after being plugged in.
- How to signal to the user that we wait for UP is env specific. OpenSK could have a small screen that says: TOUCH ME, or blink, or use a sound notification to say "Press the button!" (I hope nobody ever builds that.)
- How long and how to communicate the status back to the client is CTAP.
- Sending the resulting packet, however, is Env specific.

Applying this to the PR:
- `check_user_presence` in ctap/mod.rs is correct.
- `wait_for_user_presence_with_timeout` is the env call for button check + LEDs
- `send_or_recv_with_timeout` is the HID packet send + receive

First observation, why do both have a timeout? I think we are unintentionally changing the behavior here.
Second, at least the `quick_check` seems redundant. `check_init` and `check_complete` could be useful, but might also be avoidable with a different design.

I think getting this API right is good for us long term. We can try to get it down in one PR, but we'd need a few iterations to find something fitting. I'll have to play around with the API a bit and come back with ideas. Those are my thoughts so far.

ia0 commented:
> First observation, why do both have a timeout? I think we are unintentionally changing the behavior here.

Right, we probably don't want UP to wait with a timeout. The initial API I had in mind was:

```rust
fn check_user_presence(&self) -> bool {
	self.ext.user_presence().reset();
	for _ in 0..300 {
		if self.ext.user_presence().check() {
			return true;
		}
		self.ext.usb().send(packet());
		self.ext.time().wait(100ms);
	}
}
```

So the differences with this PR would be:
- `wait_with_timeout` doesn't wait and just returns the current status (stored in the UP object)
- no `complete` (but that's probably an oversight)

kaczmarczyck commented:
Ideally, we start immediately with
- blink
- button check
- keepalive packet
and don't simply wait 100 ms, but cancel the wait on touch. The maximum 100 ms delay is probably not noticable though. Also ideally, we don't wait 100 ms, but subtract the time the computation took so far. We'd have to measure if it's worth it, if this value is always close to zero, we can keep it simple.

Your implementation is more like `check_user_presence_hid`, but I assume you left out the boring transport details. The `self.ext.usb().send(packet());` needs to have a condition if it received a `CANCEL` to break the loop. Would we register the button callbacks in `reset` / `init` and `self.ext.user_presence().check()` simply checks? If so, we could also busy loop with a check for `self.ext.user_presence().check()`?

```
fn check_user_presence_hid(&self, hid_channel) -> Result<(), Ctap2StatusCode> {
	self.ext.user_presence().reset();
	for _ in 0..300 {
		if self.ext.user_presence().check() {
			return true;
		}
		hid_channel.send(packet());
		timer = self.ext.timer(100 ms);
		while (!timer.done() && !self.ext.user_presence().check()) {}
	}
	self.ext.user_presence().complete();
}
```
LEDs could check their timer during `ext.user_presence`.

egor-duda committed:
Move CtapHidChannel and UserPresence traits, with their accompanying
types to separate API mods. Remove Default implementations of methods
in these traits, to keep all implementation details inside of concrete
Env types.

Rename methods in UserPresence trait, for better readability.

Remove duplicate code for finding appropriate HID channel for given
transport.

Rework check_user_presence() function so that there's no more need for
quick_check() method in UserPresence trait. To short-circuit user
presence check, Env implementation may use wait_with_timeout() method.

egor-duda commented:
> First, let me start with a few goals of what we should be able to support:
> * A user presence check over HID waits for a button touch while sending HID packets and blinking LEDs.
> * A user presence check over HID during tests pretends that the button was pressed immediately.
> * The CTAP library skips user presence checks since the transport of incoming packets was NFC.

> Therefore, what is CTAP, what is Env, what is transport specific?
> * Checking UP is a CTAP function, behavior depends on transport.
> * How to receive UP over HID is Env specific. You could have an OpenSK that waits for a button, or one that doesn't even have buttons and accepts everything in the first 30 seconds after being plugged in.
> * How to signal to the user that we wait for UP is env specific. OpenSK could have a small screen that says: TOUCH ME, or blink, or use a sound notification to say "Press the button!" (I hope nobody ever builds that.)
> * How long and how to communicate the status back to the client is CTAP.
> * Sending the resulting packet, however, is Env specific.

I agree.

> Applying this to the PR:
> * `check_user_presence` in ctap/mod.rs is correct.
> * `wait_for_user_presence_with_timeout` is the env call for button check + LEDs
> * `send_or_recv_with_timeout` is the HID packet send + receive

> First observation, why do both have a timeout? I think we are unintentionally changing the behavior here.

Hmm. I don't think so. Current implementation in `develop` branch also seem to have 2 timeouts. First it waits for button touch in check_user_presence in lines
```
        // Wait for a button touch or an alarm.
        libtock_drivers::util::yieldk_for(|| button_touched.get() || keepalive_expired.get());
```
and then, if this timeout expires without user pressing the button, waits for I/O completion in send_keepalive_up_needed(), which calls usb_ctap_hid::send_or_recv_with_timeout and waits a signal from OS that I/O has been completed or I/O timeout expired. There's special TODO comment about adjusting timeouts depending of time required to perform previous wait, so that sum of two wait times was constant causing LEDs to blink in more regular pattern.

> Second, at least the `quick_check` seems redundant. `check_init` and `check_complete` could be useful, but might also be avoidable with a different design.

That function was added to be able to pretend the button was pressed immediately. `check_user_presence` function always tries to send keepalive packet first, no matter what. In TestEnv this is impossible, as it can't send/receive any packets (yet).
I was thinking of two possibilities: either check_init can return some special kind of status, meaning "skip all checking, return this result to client immediately", or some kind of special short-circuit method which allows Env to indicate that main loop in check_user_presence should be skipped altogether, to avoid any attempts to send keepalive packets.

> I think getting this API right is good for us long term. We can try to get it down in one PR, but we'd need a few iterations to find something fitting. I'll have to play around with the API a bit and come back with ideas. Those are my thoughts so far.

In this PR I've tried to keep current UP logic as close as possible to that of current implementation. Although, I understand that it may also be a good occasion to rewrite it somewhat, along with refactoring into CTAP-specific, env-specific, and transport-specific parts. 

So, for example, main cycle in check_user_presence may wait simultaneously on all three evens -- user button press + packet I/O completion + alarm expiry, instead of current implementation, which waits for button press and I/O completion in sequence.

kaczmarczyck commented:
> Hmm. I don't think so. Current implementation in `develop` branch also seem to have 2 timeouts. First it waits for button touch in check_user_presence in lines

Oh right, `send_keepalive_up_needed` is actually after the first timeout, I missed that.
 
> > Second, at least the `quick_check` seems redundant. `check_init` and `check_complete` could be useful, but might also be avoidable with a different design.
> 
> That function was added to be able to pretend the button was pressed immediately. `check_user_presence` function always tries to send keepalive packet first, no matter what. In TestEnv this is impossible, as it can't send/receive any packets (yet). I was thinking of two possibilities: either check_init can return some special kind of status, meaning "skip all checking, return this result to client immediately", or some kind of special short-circuit method which allows Env to indicate that main loop in check_user_presence should be skipped altogether, to avoid any attempts to send keepalive packets.

About `quick_check`, if we reorder instructions like in the above algorithm sketch, it shouldn't be necessary? Alternatively, since we will need a test implementation of `send_or_recv_with_timeout`, can't we just have it return `SendOrRecvStatus::Sent` and the check passes immediately after? Your code currently has a comment to extend the test env implementation, but I don't think we need more than that right now, as it is only used in user presence checks.

> In this PR I've tried to keep current UP logic as close as possible to that of current implementation. Although, I understand that it may also be a good occasion to rewrite it somewhat, along with refactoring into CTAP-specific, env-specific, and transport-specific parts.
> 
> So, for example, main cycle in check_user_presence may wait simultaneously on all three evens -- user button press + packet I/O completion + alarm expiry, instead of current implementation, which waits for button press and I/O completion in sequence.

No that's fine, we can keep functionality the same. Let's make this PR about splitting Env and Ctap correctly then!

egor-duda commented:
```
> fn check_user_presence_hid(&self, hid_channel) -> Result<(), Ctap2StatusCode> {
> 	self.ext.user_presence().reset();
> 	for _ in 0..300 {
> 		if self.ext.user_presence().check() {
> 			return true;
> 		}
> 		hid_channel.send(packet());
> 		timer = self.ext.timer(100 ms);
> 		while (!timer.done() && !self.ext.user_presence().check()) {
```
We should probably also call some kind of `self.ext.yeild()` here?
```
>              }
> 	}
> 	self.ext.user_presence().complete();
> }
```
So, with this algorithm, Env should expose the following functions: UP::reset(), UP::check(), UP::complete(), SYS::timer() and SYS::yield()?

BTW, how about this?
```
self.ext.user_presence().reset();
let result = ...;
for i in 0..=300 {
    result = self.ext.user_presence().wait_with_timeout(if i == 0 { 0 } else { 100ms });
    if result != TIMEOUT { break; }
    result = send_keepalive();
    if result.is_err() || result == CANCELED { break; }
}
self.ext.user_presence().complete();
convert_result_to_ctap2_status_code(result)
```
In this case TestEnv may return whatever it wanted to the first wait_with_timeout() call, and don't try to send anything.

kaczmarczyck commented:
Not sure if we need to yield, or Tock can do that for us. When in doubt, let's keep it simple for now. Your last proposal looks reasonable and close in functionality to what we are doing.

My guess is that we will need more iterations on this once we add more transports. So quick warning that I think this is a hard design problem I might change my mind some time in the future. :)

ia0 commented:
Yes, the last proposal looks good to me too. It hides timeout concerns in env impl which permits to avoid dealing with yield in ctap (we might eventually need it though).

egor-duda committed:
Fix button press wait with zero timeout for TockEnv

egor-duda committed:
Fix formatting

kaczmarczyck commented on src/api/user_presence.rs:
```diff
@@ -0,0 +1,42 @@
+
+pub trait UserPresence {
+    /// Called at the beginning of user presence checking process.
+    fn check_init(&mut self, _channel: Channel);
```
Do these names need underscores? Maybe I'm missing something?

egor-duda replied:
Fixed

kaczmarczyck commented on src/ctap/mod.rs:
```diff
@@ -244,6 +256,118 @@ fn verify_signature(
+        Ok(UserPresenceStatus::Timeout) => Err(Ctap2StatusCode::CTAP2_ERR_USER_ACTION_TIMEOUT),
+        Ok(UserPresenceStatus::Declined) => Err(Ctap2StatusCode::CTAP2_ERR_OPERATION_DENIED),
+        Ok(UserPresenceStatus::Confirmed) => Ok(()),
+        Ok(UserPresenceStatus::Canceled) => Err(Ctap2StatusCode::CTAP2_ERR_KEEPALIVE_CANCEL),
```
This looks like `UserPresenceResult` could also be a `Result<(), UserPresenceError>` with `UserPresenceError` being Timeout/Declined/Canceled. Then `UserPresenceError` maps to `Ctap2StatusCode` naturally. Also `send_keepalive_up_needed` could return `UserPresenceError::Canceled`.

egor-duda replied:
Done

kaczmarczyck commented on src/ctap/mod.rs:
```diff
@@ -244,6 +256,118 @@ fn verify_signature(
+        );
+        if let Ok(UserPresenceStatus::Timeout) = result {
+        } else {
+            break;
```
Maybe use `matches`?
```
if !matches!(result, Ok(UserPresenceStatus::Timeout)) {
    break;
}
```

egor-duda replied:
Done

kaczmarczyck commented on src/env/test/mod.rs:
```diff
@@ -66,13 +68,24 @@ fn new_storage() -> BufferStorage {
+        _timeout: CtapDuration,
+    ) -> SendOrRecvResult {
+        // TODO: Implement I/O from canned requests/responses for integration testing.
+        Err(SendOrRecvError)
```
This could actually pretend that `Sent` worked? Then we might not need the `i == 0` hack (but we still might want it).

egor-duda replied:
I wanted this implementation of sending via test channel to be fail-fast, to catch possible errors during testing (and it did help me on one occasion).

I guess if canned request/response functionality gets implemented, then `i == 0` check can be removed.

kaczmarczyck replied:
I don't understand what you mean by fail-fast and why one is better than the other on the regard?

egor-duda replied:
I caught an error while experimenting with quick_check, but now that method is gone, I agree that it's ok for TestEnv to always report as if send worked.
Fixed

kaczmarczyck commented on src/env/tock/mod.rs:
```diff
@@ -18,10 +20,46 @@ use rng256::TockRng256;
 mod storage;
 
+pub struct TockCtapHidChannel {
+    transport: Transport,
```
Why not hold a `usb_ctap_hid::UsbEndpoint` as state? Then you safe the match below.

egor-duda replied:
Done

kaczmarczyck commented on src/main.rs:
```diff
@@ -132,26 +134,25 @@ fn main() {
-                    usb_ctap_hid::SendOrRecvStatus::Timeout => {
+                let channel = transport.hid_channel(ctap.env());
+                match channel.send_or_recv_with_timeout(&mut pkt_reply, SEND_TIMEOUT) {
+                    Ok(SendOrRecvStatus::Timeout) => {
```
FYI: This change in main is not strictly necessary for the goal of this PR. But it doesn't hurt to have either.

egor-duda replied:
Oh, that's from the next change of adding HostEnv for QEMU emulation. May be left here, may be moved to that next PR

kaczmarczyck reviewed (commented):
I have a few style suggestions, otherwise looks pretty good! I'll have to try it on hardware to see how it works.

egor-duda committed:
Remove type for duration, use embedded_time::duration::Milliseconds
directly, for better readability.

Treat any unconfirmed result of user presence check as an error, which
maps more naturally to CTAP spec status codes.

Remove unneeded underscores in trait definition.

Store usb endpoint directly, in TockEnv channels, to avoid unneeded
conversions.

ia0 commented on src/api/channel.rs:
```diff
@@ -0,0 +1,34 @@
+
+pub type SendOrRecvResult = Result<SendOrRecvStatus, SendOrRecvError>;
+
+pub trait CtapHidChannel {
```
I'd say it's more readable if the trait has the same name as the module. The names could be `hid_channel` and `HidChannel` respectively for module and trait. I don't think we need the CTAP. But I think we need the HID (unless channel is an HID-specific concept). @kaczmarczyck what do you think? Is channel only HID-related? Or is it possible to have a non-HID channel?

kaczmarczyck replied:
Since it takes a `[u8; 64]`, it is indeed HID specific. For other naming issues, see my other comment. The reason I suggested "CtapHid" is because it implements a way to send CTAPHID packets. But since we are in a CTAP context, we can drop that if you like.

ia0 commented on src/api/user_presence.rs:
```diff
@@ -0,0 +1,40 @@
+pub enum UserPresenceError {
+    Declined,
+    Canceled,
+    Timeout,
```
I think it's worth adding a short documentation for each enum value.

egor-duda replied:
Done

ia0 commented on src/api/user_presence.rs:
```diff
@@ -0,0 +1,40 @@
+// limitations under the License.
+
+use crate::clock::ClockInt;
+use crate::ctap::Channel;
```
I don't think `api` (and `env`) should depend on `ctap`. It should ideally be the opposite. Should we move those definitions (`Channel` and `ChannelID`) to `api`? In which case, I would call the `send_or_recv` trait `transport` or `usb_hid`.

egor-duda replied:
Hmm, I was looking at customization.rs and concluded that it's ok to `use crate::ctap::...` in `api`.
If that's not advisable, then it's certainly possible to move `Channel` and `ChannelID` to `api` module.
If I understand correctly, new variants like `NFC` and `BLE` are supposed to be added to `Channel` enum in the future. In this case, it would be logical to rename `ChannelID` to something like `HidChannelID`, as that type is HID-specific.

kaczmarczyck replied:
It is technically already a `hid::ChannelID`, but we import it in a way that drops `hid`. If we move it, renaming is good. But since there is precedence, that's more of a beautification we can do now or later.

However, this made me think of a bigger naming concern I have (that is easy to fix as I see it). When I designed `Transport` and `Channel`, I intended `Transport` to be the external concept of "a physical connection" and `Channel` should be "a logical connection to one client". `Transport` is an external notion, `Channel` only known internally to the library. Users don't have to understand how we use channel IDs and put them into packets. So your API should never use `Channel`, but always `Transport`. The one `Channel` I found that is left is actually unused in `TockEnv` and I think can be removed. The `TestEnv` usage seems to be trivial. The button check function does not need to know what channel it is checking for.

On top of that, there are 2 things that are called channel now: My `Channel`, and the `CtapHidChannel` that is something else entirely.
- It is something per `Transport`, not per `Channel`.
- It's not a pure data container like `Channel` and `Transport`.
May I therefore suggest that we rename it to "HidConnection" or any other similar word you like that is not "Channel"?

I know I notice that a bit late, but I noticed this is a bit confusing as is. So the 2 action items:
- Remove `Channel` from this API, it's not needed.
- Rename `CtapHidChannel`.

ia0 replied:
> I was looking at customization.rs and concluded that it's ok to use crate::ctap::... in api

Yes, we should ideally move those enums out of ctap into api.

kaczmarczyck replied:
That's for a future PR then! The `Channel` dependency is gone.

ia0 commented on src/api/user_presence.rs:
```diff
@@ -0,0 +1,40 @@
+pub type UserPresenceResult = Result<(), UserPresenceError>;
+
+pub trait UserPresence {
+    /// Called at the beginning of user presence checking process.
```
nit: documentation should describe what the function does/should do, not when it is called.

```rust
/// Initializes for a new user presence check.
///
/// Must eventually be followed by a call to `Self::check_complete`.
```

egor-duda replied:
Done

ia0 commented on src/api/user_presence.rs:
```diff
@@ -0,0 +1,40 @@
+    /// Called at the beginning of user presence checking process.
+    fn check_init(&mut self, channel: Channel);
+
+    /// Implements a wait for user presence confirmation or rejection.
```
nit: trait documentation should describe the function behavior, it's obvious that implementers should implement that behavior

```rust
/// Waits until user presence is confirmed, rejected, or the given timeout expires.
///
/// Must be called between calls to [`Self::check_init`] and [`Self::check_complete`].
```

egor-duda replied:
Done

egor-duda committed:
No need for separate error type for send_keepalive_up_needed()

ia0 commented on src/api/user_presence.rs:
```diff
@@ -0,0 +1,40 @@
+        timeout: Milliseconds<ClockInt>,
+    ) -> UserPresenceResult;
+
+    /// Called at the end of user presence checking process.
```
nit: same here

```rust
/// Finalizes a user presence check.
///
/// Must be called after [`Self::check_init`].
```

egor-duda replied:
Done

ia0 commented on src/api/user_presence.rs:
```diff
@@ -0,0 +1,40 @@
+    ) -> UserPresenceResult;
+
+    /// Called at the end of user presence checking process.
+    fn check_complete(&mut self, result: &UserPresenceResult);
```
I'm not sure why this takes a result as argument.

kaczmarczyck replied:
I assume the use case is something like a security key with a screen, where `check_init` displays a message like "Please touch" and `check_complete` changes it to "Success" or "Timeout". But we can also remove it now and add it later if we actually need it.

ia0 replied:
Yes, I don't think we should add things we don't need, i.e.:
- the result argument of `check_complete`
- the channel argument of `check_init`

egor-duda replied:
Done

ia0 commented on src/env/tock/mod.rs:
```diff
@@ -56,8 +99,75 @@ pub fn take_storage() -> StorageResult<TockStorage> {
+            button.enable().flex_unwrap();
+        }
+
+        let immediate_check = timeout.integer() == 0;
```
Why not short-circuit the whole function?

```rust
fn wait_with_timeout() {
    if timeout.integer() == 0 {
        return Err(Timeout);
    }
```

egor-duda replied:
Done

kaczmarczyck commented on src/ctap/mod.rs:
```diff
@@ -344,27 +344,25 @@ fn check_user_presence(env: &mut impl Env, channel: Channel) -> Result<(), Ctap2
-        Ok(UserPresenceStatus::Canceled) => Err(Ctap2StatusCode::CTAP2_ERR_KEEPALIVE_CANCEL),
-        Err(_) => panic!("Unhandled error checking for user presence"),
+        Ok(()) => Ok(()),
+        Err(UserPresenceError::Timeout) => Err(Ctap2StatusCode::CTAP2_ERR_USER_ACTION_TIMEOUT),
+        Err(UserPresenceError::Declined) => Err(Ctap2StatusCode::CTAP2_ERR_OPERATION_DENIED),
+        Err(UserPresenceError::Canceled) => Err(Ctap2StatusCode::CTAP2_ERR_KEEPALIVE_CANCEL),
```
Optional: Could move into a `From` implementation in `status_code.rs`.

egor-duda replied:
Done

egor-duda committed:
Document UserPresence trait and types.

Remove unused parameters in UserPresence trait's methods.

Add conversion function from UserPresence errors to Ctap2 status codes.

Do not check button status when tock user presence wait is called with
zero timeout.

egor-duda committed:
Make test environment always report success sending data

kaczmarczyck commented on src/env/tock/mod.rs:
```diff
@@ -56,8 +99,75 @@ pub fn take_storage() -> StorageResult<TockStorage> {
+    }
+    fn wait_with_timeout(
+        &mut self,
+        _channel: Channel,
```
Unused, so shouldn't be here?

egor-duda committed:
Rename CtapHidChannel to HidConnection, rename *_hid_channel ->
*_hid_connection, for clarity. Use "Channel" to refer to the logical
connection from authenticator to one client, and use "Connection" to
refer to physical connection of authenticator to platform, on which
clients run.

Remove channel parameter from user presence API, it's not needed.

kaczmarczyck reviewed (dismissed):
Only one comment is open. @ia0 please check and merge + submit if you are happy.

ia0 commented on src/api/user_presence.rs:
```diff
@@ -13,28 +13,40 @@
+    ///
+    /// Must be called between calls to [`Self::check_init`] and [`Self::check_complete`].
+    ///
+    /// # Errors
```
nit: I would not repeat the errors here since we have them specified in the enum already. You can remove this section.

egor-duda replied:
Done

ia0 commented on src/api/user_presence.rs:
```diff
@@ -13,28 +13,40 @@
     Declined,
+    // User presence check was canceled by User Agent.
     Canceled,
+    // User presence check timed out.
```
You can use documentation here (`///`) instead of just comments (`//`).

egor-duda replied:
Done

ia0 reviewed (dismissed):
LGTM modulo some documentation remarks.

egor-duda committed:
Remove duplicate comments.

kaczmarczyck committed:
Merge branch 'develop' into user-presence-refactor
